### DISCUSSION TITLE
Add More Password Policies with an Extension?

### DISCUSSION THREAD

**Discussion by lrodriguez14** [keycloak/keycloak#13874] [Q&A]
Hello,

I have been checking the Password Policies (https://www.keycloak.org/docs/latest/server_admin/index.html#password-policy-types), and I want to know if it is possible to add more policies using an extension? It is because I need a password policy when the password is changed that the user should not repeat a number of characters from the last password. So we avoid cases like old password: **Puppy1** new password **Puppy2** and the `recently used policy` does not cover these cases.

Any help would be helpful. Also this could be an idea for an improvement for keycloak password policies.

Thanks,
-Leonardo R.

**Comment by mrpatrickpilch** [comment]
@lrodriguez14 it functions like most of the other SPIs. You create a `org.keycloak.policy.PasswordPolicyProviderFactory` with a corresponding `org.keycloak.policy.PasswordPolicyProvider` and then make sure to include a `META-INF/services/org.keycloak.policy.PasswordPolicyProviderFactory` file that contains the fully qualified class name of your provider factory. 

Sounds like you could base yours off of the existing `org.keycloak.policy.HistoryPasswordPolicyProvider`.
